zerver/middleware.py

```diff
@@ -250,6 +252,13 @@ def process_request(self, request: HttpRequest) -> None:
             # Avoid re-initializing request._log_data if it's already there.
             return
 
+        def process_user_agent() -> None:
+            request.client_name = get_client_name(request, is_browser_view=True)
+            request.client_version = None
+            if request.client_name.startswith("Zulip"):
+                request.client_version = parse_user_agent(request.META["HTTP_USER_AGENT"])["version"]
```

Comment:
This should be an independent function taking the `request` as its argument; defining a function inside another function means we do work to create the function every time `process_request` is called.

Reply:
Made it an independent method of class LogRequests.